File: solutions/20_valid_parentheses_easy.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#
# time: O(n)
# space: O(n)
# spent: 16

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', '}': '{', ']': '['}
        for char in s:
            if char in mapping:
                top = stack.pop() if stack else '#'
                if mapping[char] != top:
                    return False
            else:
                stack.append(char)
        return not stack
# @lc code=end

File: solutions/20_valid_parentheses_easy.c
//
// @lc app=leetcode id=20 lang=c
//
// [20] Valid Parentheses
//
// time: O(n)
// space: O(n)
// spent: 20
// runtime: 0ms
// memory: 8.1MB

// @lc code=start
bool isValid(char* s) {
    int len = strlen(s);
    char stack[len];
    int top = -1;
    
    for (int i = 0; i < len; i++) {
        if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
            stack[++top] = s[i];
        } else {
            if (top == -1) return false;
            if (s[i] == ')' && stack[top] != '(') return false;
            if (s[i] == '}' && stack[top] != '{') return false;
            if (s[i] == ']' && stack[top] != '[') return false;
            top--;
        }
    }
    return top == -1;
    sdfsdf
    sadasdfa
    asdfasd
}
// @lc code=end